### DISCUSSION TITLE
Ability to override/tweak `cScaleToleranceSq`

### DISCUSSION THREAD

**Discussion by mihe** [jrouwe/JoltPhysics#1126] [Ideas]
In this [latest beta release](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.13.0-beta1) of Godot Jolt I decided to include some error checking for the scaling of shapes/bodies, since I've seen a couple of instances of people scaling things in odd ways (especially compound shapes) and not realizing that this leads to problems.

I implemented this using your `IsValidScale` methods, which seemed to work great for the most part. However, I've now received reports where people have built up slight numerical inaccuracies in their transforms and ending up with scenarios like a scaling of `(0.999960, 1.000000, 0.999960)` for a capsule shape. Obviously this is technically incorrect, and can seemingly be remedied with some orthonormalization on the user's side of things, but practically speaking I'm assuming that this small of an error won't actually result in much of a problem/inaccuracy for the simulation?

As such, could the tolerance perhaps be added as a parameter to `IsValidScale` instead, which would default to `cScaleToleranceSq`, so that one could use something a bit less aggressive than the current `1.0e-8f`? I was thinking maybe something as liberal as `1.0e-4f` even.

EDIT: I guess this technically also involves the `IsNearZero()` tolerance of `1.0e-12f` in the default `Shape::IsValidScale`, but I'm not sure that one is as useful to tweak.

**Comment by jrouwe** [comment]
Hello,

I could add a tolerance to `Shape::IsValidScale`, however this function is used all over the place (search for `JPH_ASSERT(IsValidScale(inScale));`) so all of these would then need the same tolerance (and where do we get it from then?).

If I use your example in `CapsuleShapeTest::Initialize`:

```
JPH_ASSERT(big_capsule->IsValidScale(Vec3(0.999960f, 1.000000f, 0.999960f)));
```

then the assert doesn't trigger because the squared distance is 3.2e-9 which is smaller than 1.0e-8, so I would say that this particular value shouldn't have caused any issues.

That said, if the tolerance is too small then we could probably increase it a bit. Question is how much you need? I would say 1.0e-4 is way too much, this would be a cm on a m sphere.

**Reply by mihe** [reply]
It only just hit me that the tolerance was squared, despite it being in the variable name. I take back what I said about it being aggressive. 😅

That certainly makes this all very strange. I figured I might have just made too quick of an assumption during my "remote diagnosis" in thinking that the shape was just a plain `CapsuleShape`, and that maybe it was wrapped in a `StaticCompoundShape`, but that does not seem to be the case.

It is however likely wrapped in a `RotatedTranslatedShape` with a non-zero translation, but that wouldn't really change much, since it would still pass the `ScaleHelpers::IsUniformScale(inScale)` check [here](https://github.com/jrouwe/JoltPhysics/blob/729c8914f393b20851761fad7fa0bd32cd59d86d/Jolt/Physics/Collision/Shape/RotatedTranslatedShape.cpp#L288) and just defer to the inner shape's `IsValidScale` instead.

I'm not quite sure what to make of this. Some kind of shape arrangement involving a `CapsuleShape` caused `IsValidScale` to fail with a vector of `(0.999960, 1.000000, 0.999960)`.

**Reply by jrouwe** [reply]
I guess we need another repro then :smile:

**Reply by mihe** [reply]
So I had this reported to me again, and this time with a proper repro. It turned out that part of the issue was that I was running `IsValidScale` on a `CapsuleShape` that was already wrapped in a `ScaledShape` with that same (-ish) scaling, so the effective scaling ended up being squared (-ish).

Obviously that isn't much of an actual difference at these scales, but enough to push it over the edge as far as `IsValidScale` was concerned while still reporting a different (valid) scale in the error logs. The actual invalid scale for the `CapsuleShape` was (in this case) `(0.999929, 1.000000, 0.999929)`.

I'm torn on what to do here. I've painted myself into a bit of a corner by even supporting scaling bodies in Godot Jolt (unlike Godot Physics which discards any such scaling). This coupled with the fact that Godot doesn't do any sort of periodic orthonormalization of their matrix-based transforms (for understandable reasons) means that it's up to every user to understand that they need to do so instead, which is a tall order for a beginner-friendly engine like Godot. Without any orthonormalization at all it's pretty easy to find yourself with a deformed matrix after a couple of thousand transformations, to the point where `IsValidScale` will fail and start spamming errors.

I feel like the only real options are:

1. Ask for the threshold to be increased, to something like `1.0e-6f` perhaps, but that's just kicking the can down the road.
2. Remove the error-logging, but leave the checks in, as well as the resetting of the scale when the checks fail, even in non-editor (i.e. final release) builds. That'll certainly lead to confusion when it does kick in for proper big scaling errors though.
3. Remove the error-checks altogether and just accept that the collision results will be borked.

I'm mostly interested in catching the proper big scaling errors, like when people deliberately try to scale a sphere non-uniformly (or whatever) in hopes of creating some new type of shape, so having some sort of configurable threshold would be nice. That way I could set it to a value that would be unacceptable for Jolt in general but better than leaving out the checks altogether, say like `1.0e-4`.

**Reply by jrouwe** [reply]
I'm a bit torn on this subject.

A capsule shape only reads out the X component of the scale, so you can pass in any scale and it won't matter. The same holds for other shapes that have scaling restrictions. `Shape::GetWorldSpaceBounds` however doesn't know about this, so if you pass in a non-uniform scale for a sphere, you get a mismatch between what the broad phase and what the narrow phase sees.

So let's say I make the tolerance configurable, then at what point is the mismatch between broad and narrow phase big enough to cause issues? And will we be able to tell that this is the thing that is causing issues?

What about that you (or I in the form of a new `Shape::MakeValidScale` function) could discard inconsistencies in the scale on a per shape type basis? E.g. for spheres it would always return (scale.x, scale.x, scale.x). You could use a different tolerance here to check if the before / after scale changed too much and throw a warning. That way Jolt will always see the exact scale that it wants while it gives you the freedom to use any tolerance value to indicate when something is a 'user error'.

**Reply by mihe** [reply]
> What about that you (or I in the form of a new `Shape::MakeValidScale` function) could discard inconsistencies in the scale on a per shape type basis?

That does sound like a pretty good compromise actually. I'd be happy with something like `Shape::MakeValidScale`, ideally provided by Jolt itself.

**Reply by jrouwe** [reply]
Ok, I'll make it.

**Reply by jrouwe** [reply]
So we would get something like this: https://github.com/jrouwe/JoltPhysics/tree/feature/make_scale_valid

Warning: Compiles, but not tested at all.

**Reply by mihe** [reply]
Neat! That looks to be about what I was expecting, based on your description. I think that'll work nicely. I'll try to find some time in the next few days to try it out.

**Reply by jrouwe** [reply]
I made some more changes and added some unit tests. As far as I can see it works for everything except compound/rotated translated shapes now.

**Reply by jrouwe** [reply]
Code has been merged.
